src/core/api/api_patch.pl

```diff
@@ -12,16 +13,69 @@
 :- use_module(core(account)).
 :- use_module(core(query)).
 :- use_module(core(transaction)).
-
+:- use_module(core(api/api_document), [idlists_duplicates_toplevel/3]).
+:- use_module(core(document/apply), [apply_diff_ids_captures/7]).
 :- use_module(library(solution_sequences)).
 :- use_module(library(lists)).
 :- use_module(library(plunit)).
 :- use_module(library(option)).
+:- use_module(library(assoc)).
 
 api_patch(_System_DB, _Auth, Patch, Before, After, Options) :-
     % no auth yet.
     simple_patch(Patch,Before,After,Options).
 
+patch_id(Patch, Id) :-
+    do_or_die(
+        get_dict('@id', Patch, Id),
+        error(no_id_in_patch(Patch), _)
+    ).
+
+patch_id_pairs(Patch, Patch_And_Ids) :-
+    (   is_list(Patch)
+    ->  findall(P-Id,
+                (
+                    member(P, Patch),
+                    patch_id(Patch, Id)
+                ),
+                Patch_And_Ids)
+    ;   patch_id(Patch, Id),
+        Patch_And_Ids = [Patch-Id]
+    ).
+
+api_patch_resource(System_DB, Auth, Path, Patch, Commit_Info, Ids, Options) :-
+    resolve_descriptor_auth(read, System_DB, Auth, Path, instance, Branch_Descriptor),
+    create_context(Branch_Descriptor, Commit_Info, Context),
+    merge_options(Options, options{keep:json{'@id':true, '@type':true}}, Merged_Options),
+    empty_assoc(Empty),
+    with_transaction(
+        Context,
+        (
+            patch_id_pairs(Patch, Patch_And_Ids),
+            mapm({Context, Merged_Options}/[Patch-_,Conflict,Ids]>>(
+                     apply_diff_ids_captures(Context, Patch, Conflict, Ids, Merged_Options)
+                 ),
+                 Patch_And_Ids,
+                 Conflicts,
+                 Ids_List,
+                 Empty,
+                 _Captures
```

Comment:
if we thread through captures shouldn't we be checking to make sure that all captures are bound in the end?

Unless of course there is no way to actually use captures and this was just to make the calls happy.